pkg/util/tls/tls_test.go

```diff
@@ -0,0 +1,299 @@
+package tls
+
+import (
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"io/ioutil"
+	"net"
+	"os"
+	"reflect"
+	"testing"
+	"time"
+)
+
+const (
+	updateSSLVar = "UPDATE_KNOWN_SSL_CERT"
+)
+
+func readCert(path string) (*x509.Certificate, error) {
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return ParseCert(b)
+}
+
+func writeCert(path string, cert *x509.Certificate) error {
+	b, err := CertAsBytes(cert)
+	if err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(path, b, 0666)
+}
+
+func TestNewPrivateKey(t *testing.T) {
+	key, err := NewPrivateKey()
+	if err != nil {
+		t.Error(err)
+	}
+	if key.Validate() != nil {
+		t.Error(err)
+	}
+	if key.N.BitLen() < 2048 {
+		t.Errorf("insecure key length detected: %d", key.N.BitLen())
+	}
+}
+
+func TestNewCA(t *testing.T) {
+	path := "./testdata/known_good_certCA.pem"
+
+	key, cert, err := NewCA("dummy-test-certificate.local")
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = key.Validate()
+	if err != nil {
+		t.Error(err)
+	}
+	if os.Getenv(updateSSLVar) == "true" {
+		err = writeCert(path, cert)
+		if err != nil {
+			t.Error(err)
+		}
+	}
+	goodCert, err := readCert(path)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for _, c := range []*x509.Certificate{cert, goodCert} {
```

Comment:
Just to make sure I understand this test case: we're zeroing out all the info below, then checking to see if the certs match?